File: leetcode/1-100/71-simplify-path/bench/simplify.py
# Benchmark workload — Python mirror of simplify.kara.
#
# Same N=8 inputs, K=1_000_000 iters cycled by k % N. Sink is the
# sum of the simplified-output string lengths (i64), matching the
# kara/rust/c/go mirrors so all five impls print the same number.


def simplify(s: str) -> str:
    n = len(s)
    starts: list[int] = []
    ends: list[int] = []

    i = 0
    while i < n:
        while i < n and s[i] == '/':
            i += 1
        if i >= n:
            break
        j = i
        while j < n and s[j] != '/':
            j += 1
        length = j - i

        is_dot    = length == 1 and s[i] == '.'
        is_dotdot = length == 2 and s[i] == '.' and s[i + 1] == '.'

        if is_dot:
            pass
        elif is_dotdot:
            if starts:
                starts.pop()
                ends.pop()
        else:
            starts.append(i)
            ends.append(j)
        i = j

    if not starts:
        return "/"
    parts = [s[a:b] for a, b in zip(starts, ends)]
    return "/" + "/".join(parts)


def main() -> None:
    inputs = [
        "/home/",
        "/home/user/Documents/../Pictures",
        "/.../a/../b/c/../d/./",
        "/a/b/c/../..",
        "/a//b////c/d//././/..",
        "/abc_123",
        "/a/b/../c/../../d",
        "/...hidden",
    ]
    n = len(inputs)
    k_iters = 1_000_000

    total = 0
    for k in range(k_iters):
        total += len(simplify(inputs[k % n]))
    print(total)


if __name__ == "__main__":
    main()

File: leetcode/1-100/71-simplify-path/bench/simplify.rs
// Benchmark workload — Rust mirror of simplify.kara, single-threaded.
//
// Same N=8 inputs, K=1_000_000 iters cycled by k % N. Sink is the
// sum of the simplified-output string lengths (i64). Output prints
// match the kara/c/go/py mirrors so all five impls print the same
// number before timing.

fn simplify(s: &str) -> String {
    let cs: Vec<char> = s.chars().collect();
    let n: i64 = cs.len() as i64;

    let mut starts: Vec<i64> = Vec::new();
    let mut ends: Vec<i64> = Vec::new();

    let mut i: i64 = 0;
    while i < n {
        while i < n && cs[i as usize] == '/' {
            i += 1;
        }
        if i >= n {
            break;
        }
        let mut j: i64 = i;
        while j < n && cs[j as usize] != '/' {
            j += 1;
        }
        let len = j - i;

        let is_dot = len == 1 && cs[i as usize] == '.';
        let is_dotdot = len == 2 && cs[i as usize] == '.' && cs[(i + 1) as usize] == '.';

        if is_dot {
        } else if is_dotdot {
            starts.pop();
            ends.pop();
        } else {
            starts.push(i);
            ends.push(j);
        }
        i = j;
    }

    if starts.is_empty() {
        return String::from("/");
    }

    let mut out = String::new();
    for k in 0..starts.len() {
        out.push('/');
        let a = starts[k] as usize;
        let b = ends[k] as usize;
        for p in a..b {
            out.push(cs[p]);
        }
    }
    out
}

fn main() {
    let inputs: [&str; 8] = [
        "/home/",
        "/home/user/Documents/../Pictures",
        "/.../a/../b/c/../d/./",
        "/a/b/c/../..",
        "/a//b////c/d//././/..",
        "/abc_123",
        "/a/b/../c/../../d",
        "/...hidden",
    ];
    let n: i64 = inputs.len() as i64;
    let k_iters: i64 = 1_000_000;

    let mut sum: i64 = 0;
    let mut k: i64 = 0;
    while k < k_iters {
        let idx = (k % n) as usize;
        let r = simplify(inputs[idx]);
        sum += r.len() as i64;
        k += 1;
    }
    println!("{}", sum);
}
